Pokedex funcionando

Alfred021 opened the pull request:
Tiene unos bugs que solucione asi:
- Queria mostrar los elementos en la lista pokemon de 20 en 20, pero no pude sin mostrar una variable global, asi que configure para que te los mostrara de 500 en 500, asi que en teoria con hay dos paginas, cuando hacen click en next va a los siguientes 500 y en previous a los anteriores 500.

- El form de buscar no me funcionaba, solo me mostraba el pokemon que queria cuando ya habia clicleado un elemento de la lista, asi que coloque al principio de la inicializacion que el form estuviera escondido, y cuando le hicieran click a un elemento de la lista lo mostrara.

- La lista siempre me indica el el elemento activo como el ultimo, aunque selecciona los elementos correctamente.

![image](https://user-images.githubusercontent.com/31462831/79696894-c9576380-8255-11ea-887d-e1b4dbf2d2d9.png)

Alfred021 committed:
Pokedex funcionando

Tiene unos bugs que solucione asi:
- Queria mostrar los elementos en la lista pokemon de 20 en 20, pero no pude sin mostrar una variable global, asi que configure para que te los mostrara de 500 en 500, asi que en teoria con hay dos paginas, cuando hacen click en next va a los siguientes 500 y en previous a los anteriores 500.

- El form de buscar no me funcionaba, solo me mostraba el pokemon que queria cuando ya habia clicleado un elemento de la lista, asi que coloque al principio de la inicializacion que el form estuviera escondido, y cuando le hicieran click a un elemento de la lista lo mostrara:

- La lista siempre me tira el el elemento activo como el ultimo, aunque selecciona los elementos correctamente

Alfred021 committed:
Instalado Eslint, y error de elemento activo corregido

Eslint instalado como dev dependencies, corregidos errores de tabulacion en el codigo con eslint y solucionado el problema del elemento activo de la lista, ahora se marca correctamente.

fsodano commented on src/index.js:
```diff
@@ -0,0 +1,164 @@
+    }
+  });
+
+  initialUrl = `https://pokeapi.co/api/v2/pokemon/?limit=${limit}&offset=${offset}`
```
a esto le falta un `const`

Alfred021 replied:
listo

fsodano commented on src/index.js:
```diff
@@ -0,0 +1,164 @@
+  initialUrl = `https://pokeapi.co/api/v2/pokemon/?limit=${limit}&offset=${offset}`
+  getPokemon(initialUrl)
+    .then((pokemons) => {
+      listOfPokemon(pokemons);
```
vos estás consultando esta URL: https://pokeapi.co/api/v2/pokemon?offset=0&limit20

Fijate que los primeros 3 elementos son:

```"count": 964,
"next": "https://pokeapi.co/api/v2/pokemon?offset=20&limit=20",
"previous": null,```

Es decir, podés usar ese "next" y "previous" para saber cuáles son los próximos 20 (o null, si no hay ninguno) y cuáles son los 20 anteriores (o null, si no hay ninguno).

Esta es la forma en lo que lo tenés que resolver :)

Alfred021 replied:
Hola, todo bien? Vos decis que cada vez que se haga click a los botones llame al siguiente next? 
Algo asi como que haga una promesa que llame a la url.json(), y en lugar de results te llame al next, y una vez que tenga la url del next llame de vuelta al results de dicho link?

Alfred021 replied:
`let next;
let previous;

export function updateButtons(list) {
  next = list.next;
  previous = list.previous;
}

export function switchPages(list) {
  next = list.next;
  previous = list.previous;
  const $next = document.querySelector('#next');
  const $previous = document.querySelector('#previous');
  const $list = document.querySelector('#pokemon-list');

  $next.addEventListener('click', () => {
    if (next === null) {
      return;
    }
    $list.innerHTML = '';
    console.log(next);
    updatePokemonList(next);
  });

  $previous.addEventListener('click', () => {
    if (previous === null) {
      return;
    }
    if (previous !== null) {
      $list.innerHTML = '';
      console.log(previous);
      updatePokemonList(previous);
    }
    if (next === null && previous !== null) {
      $list.innerHTML = '';
      console.log(previous);
      updatePokemonList(previous);
    }
  });
}
export async function updatePokemonList(url) {
  const newPokemonResults = await getPokemon(url);
  const newPokemonList = await getPokemonList(url);
  updateButtons(newPokemonList);
  getListOfPokemon(newPokemonResults);
}
`

Esto lo solucione asi, me lo muestra corectamente ahora

fsodano commented on src/index.js:
```diff
@@ -0,0 +1,164 @@
+  $search.addEventListener('click', (event) => {
+    const pokemon = document.querySelector('input[type=search]').value.toLocaleLowerCase();
+    getPokemonInfo(`https://pokeapi.co/api/v2/pokemon/${pokemon}`)
+    event.preventDefault;
```
esto siempre al principio de la función! Si alguna de las 2 líneas anteriores falla, no va a prevenir nada.

Por otro lado, no la estás ejecutando, es decir que esta línea no está haciendo nada.

Alfred021 replied:
Esto lo cambie a esto: 

export function configureSearchBar() {
  const $search = document.querySelector('#search');
  $search.addEventListener('click', () => {
    const pokemon = document.querySelector('input[type=search]').value.toLocaleLowerCase();
    return getPokemonInfo(`https://pokeapi.co/api/v2/pokemon/${pokemon}`);
  });
  return undefined;
}

fsodano commented on src/index.js:
```diff
@@ -0,0 +1,164 @@
+function getPokemon(initialUrl) {
```
por qué initialUrl? Debería ser sólo `url`, no?

fsodano replied:
El nombre de esta función es confuso porque dice "getPokemon" singular pero es la función que en realidad se encarga de traer todos los pokemon, en este caso yo le pondría `getPokemonList` por ejemplo

Alfred021 replied:
bueno, ahora lo cambie a url

fsodano replied:
sigue siendo initialUrl

fsodano commented on src/index.js:
```diff
@@ -0,0 +1,164 @@
+    .then((resp) => resp.results)
+}
+
+function listOfPokemon(pokemons) {
```
acordate que las funciones necesitan siempre un verbo adelante, onda `getPokemonList`

Alfred021 replied:
listo, cambiado

fsodano commented on src/index.js:
```diff
@@ -0,0 +1,164 @@
+
+  pokemons.forEach((pokemon) => {
+    const name = pokemon.name
+    let url = pokemon.url
```
si esta variable no cambia dejale `const`

Alfred021 replied:
listo, eso lo cambie

fsodano commented on src/index.js:
```diff
@@ -0,0 +1,164 @@
+  });
+  $search.addEventListener('click', (event) => {
+    const pokemon = document.querySelector('input[type=search]').value.toLocaleLowerCase();
+    getPokemonInfo(`https://pokeapi.co/api/v2/pokemon/${pokemon}`)
```
El "conocimiento" de cuál es la pokeapi lo tenés disperso en varios lados.

Yo lo que tendría son 2 funciones, una que se llame getPokemonById que tome un id y el otro getPokemonList que tome un offset y un limit, que por default sean 0 y 20. Esas 2 funciones son las únicas que pueden saber que hay una API y cómo interpretar su respuesta.... todo lo demás debería llamar a esas 2 funciones si quieren saber de un pokemon o de una lista de pokeemon.

fsodano commented on src/index.js:
```diff
@@ -0,0 +1,164 @@
+  fetch(url)
+    .then((resp) => resp.json())
+    .then((pokemon) => {
+      showPokemonInfo(pokemon);
```
la función que trae la info del pokemon no debería ser la misma que muestra la info del pokemon. Por ejemplo, si tu search bar llama a esta función para obtener data de un pokemon, entonces la función de searchbar es la que tiene que agarrar esa data y pasársela a `showPokemon`.

Alfred021 committed:
Separado en Modulos

Separada la aplicacion en modulos utilizando ESM, instalado Eslint, y implementado Async/Await

Tiene un bug en a la hora de cambiar de paginas, se intenta que en lugar de ser dos paginas, sean paginas con 20 elementos c/u.

Alfred021 committed:
Corregido el problema con la paginacion.

Aunque hay un problema con la ultima pagina, hay 4 elementos y si queres volver atras te devuelve de a 4 elementos y no de a 20.

Alfred021 committed:
implementando local storage(bugs)

local storage con errores

Alfred021 committed:
implementando local storage

Se pueden guardar los datos de los pokemones, pero hay un error con las listas

Alfred021 committed:
corregido el problema de la paginacion

ahora las listas se muestran correctamente, falta implementar local storage correctamente

Alfred021 committed:
implementado local storage para la info de los pokemon

Funciona local storage para guardar y leer info de pokemon, pero aun falta que lea bien listas guardadas.

Alfred021 committed:
Ordenando el codigo un poco

fsodano commented on src/api.js:
```diff
@@ -0,0 +1,33 @@
+
+export async function configureSearchBar() {
+  const $search = document.querySelector('#search');
+  $search.addEventListener('click', () => {
+    const pokemon = document.querySelector('input[type=search]').value.toLocaleLowerCase();
+    showPokemonInfoFromService(pokemon);
+  });
```
El archivo de API sólo debería saber traer datos de la API. Este configureSearchBar va en algún módulo que sepa de UI.

fsodano commented on src/api.js:
```diff
@@ -0,0 +1,33 @@
+  const apiResp = await fetch(url);
+  const apiRespJson = await apiResp.json();
+  const pokemon = apiRespJson;
+  return showPokemonInfo(pokemon);
```
esta función sólo debería revolver el resultado de lo que trae la API, mostrar la info del pokemón le corresponde a un módulo de UI

fsodano commented on src/service.js:
```diff
@@ -0,0 +1,49 @@
+  }
+}
+
+async function getPokemonsInfo(name) {
```
`getPokemonInfo`, singular, porque entiendo que esto trae sólo 1 pokemón

fsodano commented on src/service.js:
```diff
@@ -0,0 +1,49 @@
+}
+
+export async function updatePokemonListFromService(url) {
+  const newPokemon = await getPokemonsFromList(url);
+  getListOfPokemon(newPokemon);
```
esto está complicado de seguir, cuál es la diferencia entre`getPokemonsFromList`y `getListOfPokemon`?

fsodano commented on src/index.js:
```diff
@@ -0,0 +1,154 @@
+}
+
+
+function switchPages(offset) {
```
tenés 2 de estas funciones

fsodano commented on src/pagination.js:
```diff
@@ -0,0 +1,35 @@
+  previous = url.previous;
+}
+
+export function switchPages(funtion) {
```
`funtion` es mejor nombrada como `callback`

fsodano commented on src/pagination.js:
```diff
@@ -0,0 +1,35 @@
+let next;
+let previous;
+
+export function updateButtons(url) {
```
ponerle `url` a esto es engañoso porque lo que recibe es todo el objeto en si

Alfred021 committed:
Arreglando local storage

Alfred021 committed:
Local storage funcionando correctamente en listas y pokemones

Corregido el problema con las listas, ahora local storage ubica las listas guardadas sin problema.

Alfred021 committed:
Instalando Babel, Jest, y escribiendo algunos tests con Jest

Solo los archivos index, main y api estan testeados por el momento.

Alfred021 committed:
Aplicando Clases

Aplicada una clase para la respuesta de la api en la UI.

fsodano commented on src/class.js:
```diff
@@ -0,0 +1,13 @@
+/* eslint-disable linebreak-style */
+export default class pokemonGeneralInfo {
```
las clases van en PascalCase

Ponele class Pokemon y ya :)

fsodano commented on src/class.js:
```diff
@@ -0,0 +1,13 @@
+/* eslint-disable linebreak-style */
+export default class pokemonGeneralInfo {
+  constructor(pokemon) {
+    this.name = pokemon.name.toLocaleLowerCase();
```
no puede haber ningún tipo de lógica en la entidad

fsodano commented on src/class.js:
```diff
@@ -0,0 +1,13 @@
+    this.id = pokemon.id;
+    this.weight = pokemon.weight;
+    this.height = pokemon.height;
+    this.image = pokemon.sprites.front_default;
```
si bien esto para empezar está bien, fijate que tu entidad está demasiado acoplada a la respuesta de la api. La estructura del objeto pokemon que recibís como parámetro vos estás asumiendo que es la misma que la API de pokeapi.co, lo que significa que si el día de mañana cambiás de API también vas a tener que modificar esta clase, que no está bien.

Lo ideal es que tengas otro módulo más que sea un `mapper`, que sepa convertir respuestas de la API de pokeapi.co en entidades de Pokemon de tu aplicación.

A esta clase, entonces, la podés modificar para que ya tome todos los parámetros individuales.


```
// esto es object destructuring implementado en el constructor
constructor({id, name, weight, height, image}){
   this.id = id;
   this.name = name;
   // etc.
}
```

fsodano commented on src/ui/ui.js:
```diff
@@ -0,0 +1,99 @@
+  });
+}
+
+export function showPokemonInfo(pokemon) {
```
esto ya debería tomar una entidad Pokemon

fsodano commented on src/service/service.js:
```diff
@@ -0,0 +1,33 @@
+  getPokemonList,
+} from '../api/api.js';
+
+export async function getPokemonsFromList(url) {
+  try {
+    const pokemonList = getPokemonListFromStorage(url);
+    return pokemonList;
+  } catch (e) {
+    const pokemonList = await getPokemonList(url);
+    savePokemonListOnStorage(url, pokemonList);
+    return pokemonList;
+  }
+}
```
esta función está guardando en local storage (y por lo tanto devolviendo) una lista de pokemones tal cual viene en la API. En vez de eso, debería estar guardando entidades propias de esta aplicación, de clase Pokemon.

fsodano commented on src/service/service.js:
```diff
@@ -0,0 +1,33 @@
+  }
+}
+
+export async function getPokemonsInfo(name) {
```
`getPokemon`

2 cosas:
1. el `info` está de más, qué estás obteniendo del pokemon si no es su información?
2. singular, getPokemon me da a entender que obtenés sólo 1, get Pokemons me da a entender que son varios.

fsodano commented on src/service/service.js:
```diff
@@ -0,0 +1,33 @@
+
+export async function getPokemonsInfo(name) {
+  try {
+    const pokemon = getPokemonInfoFromStorage(name);
+    return pokemon;
```
```suggestion
    return getPokemonInfoFromStorage(name);
```

fsodano commented on src/ui/pagination.js:
```diff
@@ -0,0 +1,25 @@
+      return;
+    }
+    if (nextUrl === null) {
+      callback(previousUrl = 'https://pokeapi.co/api/v2/pokemon/?offset=940&limit=20');
```
la UI no puede saber de una URL de la API.

Si mañana cambia la API ya tenés que venir acá y cambiar esto también, es decir que tu lógica está desparramada en vez de contenida en la capa de API.

fsodano commented on src/ui/ui.js:
```diff
@@ -0,0 +1,99 @@
+/* eslint-disable linebreak-style */
```
si lo necesitás en todo tu proyecto configuralo en el archivo de configuración de eslint. Si no, no veo por qué estaría especificado sólo en este archivo

fsodano commented on src/ui/ui.js:
```diff
@@ -0,0 +1,99 @@
+  const $pokemonImage = document.querySelector('#pokemon-image');
+  const $pokemonweight = document.querySelector('#pokemon-weight');
+  const $pokemonheight = document.querySelector('#pokemon-height');
+  const pokemonInfo = new pokemonGeneralInfo(pokemon);
```
como esto debería ser una entidad pokemon, no te haría falta hacer esto

fsodano commented on src/ui/ui.js:
```diff
@@ -0,0 +1,99 @@
+  const $pokemonheight = document.querySelector('#pokemon-height');
+  const pokemonInfo = new pokemonGeneralInfo(pokemon);
+
+  $pokemonName.textContent = `No.${pokemonInfo.id} ${pokemonInfo.name.toUpperCase()}`;
```
el upperCase acá está bárbaro, porque es lógica de presentación, que está en la capa de UI. 💯

fsodano commented on css/index.css:
```diff
@@ -0,0 +1,13 @@
+}
+
+#pokemon-list {
+    height: 8cm;
```
usá px, %, vh ó em. No cm.

fsodano commented on src/api/api.js:
```diff
@@ -0,0 +1,13 @@
+/* eslint-disable linebreak-style */
+
+export async function getPokemonList(url = 'https://pokeapi.co/api/v2/pokemon/') {
+  const apiResp = await fetch(url);
+  const apiRespJson = await apiResp.json();
+  return apiRespJson;
+}
```
cuando uno declara una función como async, el resultado si o si va a ser una promesa, es decir que el último await está de más!
```suggestion
export async function getPokemonList(url = 'https://pokeapi.co/api/v2/pokemon/') {
  const response = await fetch(url);
  return response.json();
}
```

fsodano commented on src/api/api.js:
```diff
@@ -0,0 +1,13 @@
+/* eslint-disable linebreak-style */
+
+export async function getPokemonList(url = 'https://pokeapi.co/api/v2/pokemon/') {
```
la base de la URL la podés poner en una constante en este archivo:

`const URL_BASE = 'https://pokeapi.co/api/v2/pokemon'`

fsodano commented on src/class.js:
```diff
@@ -0,0 +1,13 @@
+    this.weight = pokemon.weight;
+    this.height = pokemon.height;
+    this.image = pokemon.sprites.front_default;
+    this.types = pokemon.types;
+    this.skills = pokemon.abilities;
+    this.stats = pokemon.stats;
```
Estas 3 son arrays. Arrays de objetos, con una estructura predeterminada.

Es decir que necesitás 3 clases/entidades más: `PokemonType`, `PokemonAbility` y `PokemonStat` (ó sólo `Type`, `Ability` y `Stat`).

fsodano commented on src/main.js:
```diff
@@ -0,0 +1,22 @@
+} from './ui/ui.js';
+import {
+  getPokemonsFromList,
+} from './service/service.js';
+
+export default async function setUp(url) {
```
acá pedís una URL pero nunca nadie llama a esta función con una url, para qué está?

Alfred021 replied:
Si, se llama como un callback con una url, en el archivo pagination: 

`export default function switchPages(nextUrl, previousUrl, callback) {
  const $next = document.querySelector('#next');
  const $previous = document.querySelector('#previous');
  const $list = document.querySelector('#pokemon-list');

  $next.onclick = () => {
    if (nextUrl === null) {
      return;
    }
    $list.innerHTML = '';
    callback(nextUrl);
  };

  $previous.onclick = () => {
    if (previousUrl === null) {
      return;
    }
    if (nextUrl === null) {
      callback(previousUrl = 'https://pokeapi.co/api/v2/pokemon/?offset=940&limit=20');
    }
    $list.innerHTML = '';
    callback(previousUrl);
  };
}`

fsodano commented on src/ui/ui.js:
```diff
@@ -0,0 +1,99 @@
+    const pokemonInfo = await getPokemonsInfo(pokemon);
+    showPokemonInfo(pokemonInfo);
+  };
+  return undefined;
```
esto no te hace falta, por default la función sin return explícito devuelve undefined

Alfred021 committed:
Actualizando Readme

Alfred021 committed:
Actualizando Readme

Alfred021 committed:
Agregando Entidades y Mapeadores para la respuesta de la API

Agregadas entidades a la respuesta de la api, y un mapeador para interpretarlas, tambien instalado cypress.
